apps/frontend/components/RouteSearch.tsx

```diff
@@ -0,0 +1,76 @@
+"use client";
+
+import axiosInstance from "@/axios/axiosInstance";
+import React, { useEffect, useState } from "react";
+
+import { Route } from "../interfaces";
+import AvailableRouteList from "./AvailableRouteList";
+import BestRoute from "./BestRoute";
+import CountrySelector, { COUNTRY_TYPE } from "./CountrySelector";
+import ErrorMessage from "./ErrorMessage";
+
+const defaultCheapestRoute = { departurePort: "", arrivalPort: "", distance: 0 };
+
+const RouteSearch: React.FC = () => {
+  const [sourceCountry, setSourceCountry] = useState("");
+  const [destinationCountry, setDestinationCountry] = useState("");
+  const [cheapestRoute, setCheapestRoute] = useState<Route>(defaultCheapestRoute);
+  const [availableCountries, setAvailableCountries] = useState<string[]>([]);
+  const [errorMessage, setErrorMessage] = useState<string>("");
+  const [routes, setRoutes] = useState<Route[]>([]);
+  const [isShowAll, setIsShowAll] = useState<boolean>(false);
+
+  useEffect(() => {
+    const fetchAvailableCountries = async () => {
+      try {
+        const { data } = await axiosInstance.get("/port/listCountries");
+        setAvailableCountries(data);
+        setSourceCountry(data[0]);
+        setDestinationCountry(data[1]);
+      } catch (error: any) {
+        setErrorMessage(error?.message || "Something went wrong to get countries list! Please try again.");
+      }
+    };
+    fetchAvailableCountries();
+  }, []);
```

Comment:
**suggestion:** similarly on the FE, having you api call directly in a useEffect is a bit shortcut 😄 . A good approach could be to use external hook and core classes that fetch from external api, and dissociate your UI components from your fetching logic. This help a lot to test unit your components, and segment stuff. In case you need to swtich to graphql api instead of rest for example, here you will have to rewrite plenty code directly in your components, that could be extracted from here for more maintainability.